packages/runtime/container-runtime/src/test/batching.spec.ts

```diff
@@ -0,0 +1,229 @@
+		return patched as unknown as ContainerRuntime;
+	}
+
+	/**
+	 * Returns a batch of messages with the first and last message marked as batch start and end respectively.
+	 */
+	function getBatch(count: number): ISequencedDocumentMessage[] {
+		const messages: ISequencedDocumentMessage[] = [];
+		for (let i = 0; i < count; i++) {
+			messages.push({
+				type: MessageType.Operation,
+				clientId: mockClientId,
+				sequenceNumber: i,
+				minimumSequenceNumber: 1,
+				contents: {
+					type: ContainerMessageType.FluidDataStoreOp,
+					contents: {
+						address: "address",
+						contents: "dummy content",
+					},
+				},
+				clientSequenceNumber: i,
+				timestamp: Date.now(),
+			} as unknown as ISequencedDocumentMessage);
+		}
+		messages[0].metadata = { batch: true };
+		messages[messages.length - 1].metadata = { batch: false };
+		return messages;
+	}
+
+	/**
+	 * Processes the given batch. The batch is processed by pushing each message to the inbound queue and then
+	 * processing the messages in the queue.
+	 */
+	function processBatch(batch: ISequencedDocumentMessage[], cr: ContainerRuntime) {
+		// Push the messages in the inbound queue. This is done because ScheduleManager listens to the "push" event
+		// emitted by the inbound queue to do batch validations.
+		for (const batchMessage of batch) {
+			mockDeltaManager.inbound.push(batchMessage);
+		}
+
+		// Process the messages in the inbound queue.
+		// Process is called on the delta manager because ScheduleManager listens to the "op" event on delta manager
+		// as well to do validation.
+		// Process is called on the container runtime because it is the one that actually processes the messages and
+		// has its own set of validations.
+		let message = mockDeltaManager.inbound.pop();
+		while (message !== undefined) {
+			assert(message !== undefined, "Message should not be undefined");
+			mockDeltaManager.process(message);
+			cr.process(message, false /* local */);
+			message = mockDeltaManager.inbound.pop();
+		}
+	}
+
+	it("successfully processes messages that are not part of batch", async () => {
+		const cr = patchContainerRuntime(containerRuntime);
+		const messageCount = 5;
+		const batch = getBatch(messageCount);
+
+		// Remove the batch metadata essentially making the messages not part of a batch.s
+		batch[0].metadata = undefined;
+		batch[messageCount - 1].metadata = undefined;
+
+		assert.doesNotThrow(
+			() => processBatch(batch, cr),
+			"Non batch messages should be processed successfully",
+		);
+	});
+
+	it("successfully processes a batch containing ops from a single client", async () => {
+		const cr = patchContainerRuntime(containerRuntime);
+		const messageCount = 5;
+		const batch = getBatch(messageCount);
+
+		assert.doesNotThrow(
+			() => processBatch(batch, cr),
+			"Batch from a single client should be processed successfully",
+		);
+	});
+
+	it("fails processing a batch with batch end but no batch start", async () => {
```

Comment:
Another test case is 2 starts with no end in between

Reply:
See also:  "Throws on invalid batches" describe block in remoteMessageProcessor.spec.ts

Reply:
Nice. Between the tests in these 2 files, we should have enough coverage!